solana-programs/staking-bridge/programs/staking-bridge/src/lib.rs

```diff
@@ -1,15 +1,271 @@
+    check_raydium_programs,
+    check_raydium_token_accounts,
+    check_raydium_pdas,
+    swap
+};
+use crate::wormhole_utils::{
+    check_wormhole_programs,
+    check_wormhole_token_accounts,
+    check_wormhole_pdas,
+    approve_wormhole_transfer,
+    execute_wormhole_transfer
+};
 
 declare_id!("HEDM7Zg7wNVSCWpV4TF7zp6rgj44C43CXnLtpY68V7bV");
 
 #[program]
 pub mod staking_bridge {
     use super::*;
 
-    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+    /**
+     * Creates the PDA.
+     * This instruction can be called by anyone.
+     * Immediately returns successfully because Anchor handles
+     * the PDA creation via the CreatePda struct account macros.
+     */
+    pub fn create_pda(_ctx: Context<CreatePda>) -> Result<()> {
+        Ok(())
+    }
+
+    /**
+     * Verifies that correct programs, token accounts, PDAs are being used,
+     * and then swaps tokens via the Raydium AMM Program.
+     */
+    pub fn raydium_swap(
+        ctx: Context<RaydiumSwap>,
+        amount_in: u64,
+        minimum_amount_out: u64,
+        vault_nonce: u64,
+        staking_bridge_pda_bump: u8
+    ) -> Result<()> {
+        let accounts = ctx.accounts;
+        check_raydium_programs(accounts)?;
+        check_raydium_token_accounts(accounts)?;
+        check_raydium_pdas(
+            accounts,
+            vault_nonce
+        )?;
+        swap(
+            accounts,
+            amount_in,
+            minimum_amount_out,
+            staking_bridge_pda_bump
+        )?;
+        Ok(())
+    }
+
+    pub fn post_wormhole_message(
+        ctx: Context<PostWormholeMessage>,
+        nonce: u32,
+        amount: u64, // TODO: if amount greater than max amount of our PDA, do we send max or do we fail?
+        fee: u64, // TODO: do we need this? what if someone passes in a huge fee?
+        config_bump: u8,
+        wrapped_mint_bump: u8,
+        wrapped_meta_bump: u8,
+        authority_signer_bump: u8,
+        bridge_config_bump: u8,
+        emitter_bump: u8,
+        sequence_bump: u8,
+        fee_collector_bump: u8,
+        staking_bridge_pda_bump: u8,
+    ) -> Result<()> {
+        let accounts = ctx.accounts;
+        check_wormhole_programs(accounts)?;
+        check_wormhole_token_accounts(accounts)?;
+        check_wormhole_pdas(
+            accounts,
+            config_bump,
+            wrapped_mint_bump,
+            wrapped_meta_bump,
+            authority_signer_bump,
+            bridge_config_bump,
+            emitter_bump,
+            sequence_bump,
+            fee_collector_bump,
+        )?;
+        approve_wormhole_transfer(
+            accounts,
+            amount,
+            staking_bridge_pda_bump
+        )?;
+        execute_wormhole_transfer(
+            accounts,
+            nonce,
+            amount,
+            fee,
+            staking_bridge_pda_bump
+        )?;
         Ok(())
     }
 }
 
 #[derive(Accounts)]
-pub struct Initialize {}
+pub struct CreatePda<'info> {
```

Comment:
maybe we should name this something more specific, like

`CreateStakingBridgeBalancePda`

or StakingBridgeFunds?